### ISSUE TITLE
pycdf module not thread safe at all

### ISSUE THREAD

**Issue by jeandet** [spacepy/spacepy#261]
Hello,
Unless I missed something, the pycdf module seems to fail randomly when used from multiple threads, I randomly get different errors depending on if I read/write many files in //. 
When writing files I get python error and segfault when reading in //.

### Minimal example to reproduce issue:
If you play with ```thread_count``` you might get errors more or less often.
With thread_count that should always work.

```
from multiprocessing import dummy
from spacepy import pycdf as cd
from datetime import datetime 
import numpy as np
import os
def make_cdf(cdf_file_name):
    if os.path.exists(cdf_file_name):
        os.remove(cdf_file_name)
    cdf_file = cd.CDF(cdf_file_name, '')
    time = [datetime(2000, 10, 1, 1, val) for val in range(60)]
    time_epoch = [t.timestamp() for t in time]
    values = np.random.random_sample(len(time))
    cdf_file['Epoch'] = time
    cdf_file['data'] = values
    cdf_file['data'].attrs['DEPEND_0'] = 'Epoch'
    cdf_file.close()
    os.remove(cdf_file_name)
thread_count = 4
with dummy.Pool(thread_count) as p:
    p.map(make_cdf, [f"cdf_file_{i}.cdf" for i in range(thread_count)])
```

### Error message/Traceback when writing in//:

```
CDFError                                  Traceback (most recent call last)
<ipython-input-3-d62502d322c2> in <module>
     18 proc_count = 4
     19 with dummy.Pool(proc_count) as p:
---> 20     p.map(make_cdf, [f"cdf_file_{i}.cdf" for i in range(proc_count)])
     21 

/usr/lib64/python3.7/multiprocessing/pool.py in map(self, func, iterable, chunksize)
    266         in a list that is returned.
    267         '''
--> 268         return self._map_async(func, iterable, mapstar, chunksize).get()
    269 
    270     def starmap(self, func, iterable, chunksize=None):

/usr/lib64/python3.7/multiprocessing/pool.py in get(self, timeout)
    655             return self._value
    656         else:
--> 657             raise self._value
    658 
    659     def _set(self, i, obj):

/usr/lib64/python3.7/multiprocessing/pool.py in worker(inqueue, outqueue, initializer, initargs, maxtasks, wrap_exception)
    119         job, i, func, args, kwds = task
    120         try:
--> 121             result = (True, func(*args, **kwds))
    122         except Exception as e:
    123             if wrap_exception and func is not _helper_reraises_exception:

/usr/lib64/python3.7/multiprocessing/pool.py in mapstar(args)
     42 
     43 def mapstar(args):
---> 44     return list(map(*args))
     45 
     46 def starmapstar(args):

<ipython-input-3-d62502d322c2> in make_cdf(cdf_file_name)
     12     values = np.random.random_sample(len(time))
     13     cdf_file['Epoch'] = time
---> 14     cdf_file['data'] = values
     15     cdf_file['data'].attrs['DEPEND_0'] = 'Epoch'
     16     cdf_file.close()

~/.local/lib/python3.7/site-packages/spacepy/pycdf/__init__.py in __setitem__(self, name, data)
   1804                 self[name].attrs.clone(data.attrs)
   1805         else:
-> 1806             self.new(name, data)
   1807 
   1808     def __iter__(self, current = 0):

~/.local/lib/python3.7/site-packages/spacepy/pycdf/__init__.py in new(self, name, data, type, recVary, dimVarys, dims, n_elements, compress, compress_param)
   2339             new_var.compress(compress, compress_param)
   2340         if data is not None:
-> 2341             new_var[...] = data
   2342             if hasattr(data, 'attrs'):
   2343                 new_var.attrs.clone(data.attrs)

~/.local/lib/python3.7/site-packages/spacepy/pycdf/__init__.py in __setitem__(self, key, data)
   3111             hslice.select()
   3112             lib.call(const.PUT_, const.zVAR_HYPERDATA_,
-> 3113                      data.ctypes.data_as(ctypes.c_void_p))
   3114         if hslice.counts[0] < n_recs:
   3115             first_rec = hslice.starts[0] + hslice.counts[0]

~/.local/lib/python3.7/site-packages/spacepy/pycdf/__init__.py in call(self, *args, **kwargs)
    631         else:
    632             return self.check_status(self._library.CDFlib(
--> 633                 *(args + (const.NULL_, ))
    634                 ))
    635 

~/.local/lib/python3.7/site-packages/spacepy/pycdf/__init__.py in check_status(self, status, ignore)
    583             return status
    584         if status < const.CDF_WARN:
--> 585             raise CDFError(status)
    586         else:
    587             warning = CDFWarning(status)

CDFError: NO_VAR_SELECTED: A variable has not been selected.
```

### OS, Python version, and dependency version information:
Linux-5.3.15-300.fc31.x86_64-x86_64-with-fedora-31-Thirty_One
sys.version_info(major=3, minor=7, micro=5, releaselevel='final', serial=0)
numpy=1.17.4
scipy=1.2.1
matplotlib=3.1.1

### Version of SpacePy
spacepy=0.2.1 from PyPi

**Comment by jeandet** [comment]
I forgot to mention that I use cdf lib 3.7.1 from https://spdf.gsfc.nasa.gov/pub/software/cdf/dist/cdf37_1/

**Comment by jtniehof** [comment]
Pretty sure the CDF library itself isn't thread-safe (the word "thread" occurs not at all in the CDF C reference manual), because it depends on setting internal state of the library and then performing operations, and there's only one state of the library. I could potentially put a lock on the actual call method to serialize operations (the interface that pycdf presents is essentially atomic), which would probably do the job. This manifests partly because ctypes code drops the GIL so we can have two threads running simultaneously in the actual call to the library.

It *might* also be possible to load the actual library once in each thread to get independent state but I'd have to think carefully about other issues that could open.

At a minimum this needs documenting with a recommendation to avoid threading I/O.

Completely unrelated: your example would result in the Epoch variable being a float not a CDF_EPOCH/CDF_EPOCH16/CDF_TIME_TT2000. Assigning the datetime directly will result in a time-type variable.

**Comment by balarsen** [comment]
I figured I would share my couple of thoughts here for whatever it is worth:
1) Based on the underlying CDF library not being thread safe spacepy adding that seems hard to impossible, so documenting it moving on seems appropriate. 
2) Number 1 is just a kick the can down the road, I believe that as a NASA community we owe it to ourselves to advocate for the improvement of (or dropping of) the CDF back-end and capabilities. We have moved all our non-NASA data from CDF to HDF5 and I am always impressed with how much better it is (groups, compression options, pointers to other files, complex data types). Netcdf files are hdf5 under the hood now as well. All this really says is that if as a community we feel that CDF still adds value we need to write proposals or push NASA to fund a full update, it will not be cheap or sexy but a (in my feeling) dying format is no way to do business.

**Comment by jeandet** [comment]
@jtniehof thank you for your quick answer, I didn't want to do any assumption on CDF lib without reading the code (what I won't do to protect my eyes). But yeah, from outside it looks like it has some global state/buffers... What I don't really understand, is that nobody has ever made a multi-thread CDF reader/writer yet? 

So yeah a big warning in the documentation seems ok to me.

**Comment by jeandet** [comment]
@balarsen I would agree, but sadly this doesn't seems to append. For example on Solar Orbiter we use cdf, I guess JUICE will be the same. And I don't see them changing file format afterwards.

In the other hand I can't say that netCDF or HDF format have modern, clean and future proof implementations.
My understanding is that we suffer more from legacy code than features missing in X or Y format.

**Comment by drsteve** [comment]
Two quick comments:

1. On the issue of loading the CDF library once in each thread to have independent states... This used to be extremely problematic, due to CDF's use of temporary files. The temporary filenames used the time as a seed, so clashes occurred when running multiple independent processes (discovered and "fixed" in the early days of the MMS mission). I believe the process id has now been added to the seed, so this is far less likely to happen. Of course, there may still be issues.

2. While nothing is future-proof, in our current ecosystem I'd say HDF5 is closest due to its support from the HPC community. As far as I'm aware there shouldn't be an issue with parallel use of that library as long as individual files aren't being written to by multiple threads... HDF5 does have a parallel implementation, and that is supported by more recent versions of h5py (which is the library we use to provide access in datamodel). There are moves at NASA SPDF to also support netCDF4, but in the short term CDF isn't going anywhere.

3. netCDF4 is really a superset of HDF5. It adds some arguably good things, as well as a lot of legacy support for prior incarnations of netCDF. netCDF4 goes using extended features are not directly readable as HDF5 (enums in negative are an example I've got). My personal view, having worked with it for several projects, is that I'm in favor of sticking with HDF5 and specifying a strong metadata model. But either would be a better choice than NASA CDF, assuming we get to make a choice.

Finally, my thought on how to proceed here is to first document and discourage threading when using CDF. Multiprocessing, where each new process loads the CDF library is likely to be the best way to work in parallel. We should probably look into that and give an example of a clean/safe way of doing that (noting that -however unlikely - it's still technically possible to get unexpected behavior due to the use of temporary files).

I'll ask about future plans for CDF next time I have a telecon with SPDF.

**Comment by jtniehof** [comment]
Testing a lock on the library is fairly straightforward so I'll take a look and see what impact it has on performance.

As far as the future of CDF+ISTP+NASA etc., bigger issue and we should probably also include Jeremy Faden at least. Maybe kick some ideas around this weekend, depending on if it's an A or B Friday this week.

**Comment by jeandet** [comment]
I started a pure C++ impl to get rid off this limitations, I'll also push wheels on PyPi once I get CDF reading complete.
[Here](https://github.com/SciQLop/CDFpp) is the code, any contributions, help, tests, feedbacks are welcome :).
It'll be wrapped with Pybind11, I hope it'll be faster than [cdflib](https://github.com/MAVENSDC/cdflib).
I'll also add the ability to load in memory files which avoids to use fsram/tmp or whatever when you want to load a file from network without writing it on disk.
One other main difference is that it doesn't operate in place in files, instead it loads in memory a repr of the file. This implies that loading a file and writing it back will most likely change the file.

**Comment by balarsen** [comment]
@jtniehof, all of a sudden I am finding myself overly paranoid but I will ask anyway. This 

> ... because it depends on setting internal state of the library and then performing operations...

Am I sure to be safe from process to process? Assuming different files. Like can I read 10 CDF files in parallel and be good to go? I think so but feeling paranoid...

**Comment by drsteve** [comment]
@balarsen See my note about updates of the CDF library to include the process ID. I believe that separate processes should be fine. Technically there's a small chance of clashes still, but in practice it's a vanishingly small chance. Most importantly, I haven't heard of any further issues from the MMS SDC or from the instrument teams, and they're using CDF this way (fairly heavily).

**Comment by jtniehof** [comment]
@balarsen, you should be fine. There are at least two issues:

1. @drsteve has mentioned the potential for clashes in temporary file names between processes, which should be taken care of.
2. The interface "remembers" what a process is working on (the APi works like: "I want this CDF. I want this variable in the current CDF. I want to do this to the current variable.") and if there's a thread switch between making a selection and acting it'd be bad.
3. I'm just going to assume there are other things about the CDF library that are re-entrant.

Putting a lock as I'm planning to do would handle 2 and 3, as would working in separate processes.

**Comment by jtniehof** [comment]
It's also worth noting that it's possible to open the same CDF in two different processes and the results are fairly weird--deleting a variable in one process and closing the CDF still leaves the variable visible in the other process. I think that's pretty firmly in don't-do-that territory and will put on some documentation for that.

**Comment by berniegsfc** [comment]
For the last few years, I think cdf has used mkstemp to create its temporary files so I don't think temporary files will be a problem (unless someone is using an old version of cdf).  But I think there are plenty of other problems keeping the code from being thread-safe.  The Java JNI code ensures the C library is never called simultaneously from more than 1 thread (which creates a big bottleneck in multi-threaded code).

**Comment by jtniehof** [comment]
Thanks Bernie. That was pretty much my plan. I need to actually get to it :)

**Comment by berniegsfc** [comment]
BTW, if you ever need any data from cdaweb in the spdm, you might try [cdasws](https://pypi.org/project/cdasws/) or better at the moment [cdasws](https://test.pypi.org/project/cdasws/).

**Comment by jtniehof** [comment]
Sweet! We can definitely document that. Do you have a Sphinx doc output so we can cross-link?

You might also be interested in the VarBundle class that will be in the next release and makes it easy to grab variable + dependencies. It only outputs to CDF right now but I'm going to extend it to SpaceData.

**Comment by berniegsfc** [comment]
I forgot to mention that get_data() even works for the netcdf (ICON, GOLD, and a few GOES) datasets in cdaweb.  And that [cdasws](https://anaconda.org/spdf.gsfc.nasa.gov/cdasws) is available on anaconda.org.  Is SpacePy going to be available there?  PySpedas needed cdasws on anaconda.org.

Sphinx doc looks like a lot of trouble given that my current documentation is markdown and numpy docstring (requiring m2r and napoleon sphinx extensions).  Note that the readme/description on pypi.org and anaconda.org has links to the [API](https://cdaweb.sci.gsfc.nasa.gov/WebServices/REST/py/cdasws/index.html) and [FAQ](https://cdaweb.sci.gsfc.nasa.gov/WebServices/REST/py/FAQ.html).     Do you need more than that?  Is sphinx really worth it?

VarBundle does sound interesting.  I'd also be interested in some little istp attribute functions but I wasn't sure how to add them.  For example, get_labels that looked for LABLAXIS and LABL_PTR_n attributes and returns the correct string values.

**Comment by jtniehof** [comment]
Note that with the eventual removal of the GIL, we probably do need to implement this locking.
